tests/puzzle/test_solver.py

```diff
@@ -1183,6 +1184,65 @@ def test_solver_with_dependency_and_prerelease_sub_dependencies(
     )
 
 
+def test_solver_with_dependency_and_prerelease_sub_dependencies_increasing_constraints(
+    solver: Solver, repo: Repository, package: ProjectPackage
+) -> None:
+    """Regression test to ensure the solver eventually uses pre-release
+    dependencies if the package is progressively constrained enough.
+
+    This is different from test_solver_with_dependency_and_prerelease_sub_dependencies
+    above because it also has a wildcard dependency on B at the root level.
+    This causes the solver to first narrow B's candidate versions down to
+    {0.9.0} at an early level, then eventually down to the empty set once A's
+    dependencies are processed at a later level.
+
+    Once the candidate version set is narrowed down to the empty set, the
+    solver should re-evaluate available candidate versions from the source, but
+    include pre-release versions this time as there are no other options.
+    """
```

Comment:
Sorry for all the verbosity—I was having a tough time  finding words to clearly describe the situation I'm trying to test. Open to suggestions to improve this!

Reply:
Many of our tests are not documented well and sometimes you even have to guess the objective of the test. IMO, the description is fine here considering it's a quite complex use case.